junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java

```diff
@@ -1479,12 +1561,97 @@ private static void failArraysNotEqual(Object expected, Object actual, Deque<Int
 		return result;
 	}
 
-	private static String formatIndexes(Deque<Integer> indexes) {
-		if (indexes == null || indexes.isEmpty()) {
-			return "";
+	// --- assertIterableEquals helpers ----------------------------------
+
+	private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == actual) {
+			return;
 		}
-		String indexesString = indexes.stream().map(Object::toString).collect(joining("][", "[", "]"));
-		return " at index " + indexesString;
+		assertIterablesNotNull(expected, actual, indexes, messageSupplier);
+
+		Iterator<?> expectedIterator = expected.iterator();
+		Iterator<?> actualIterator = actual.iterator();
+
+		int processed = 0;
+		while (expectedIterator.hasNext() && actualIterator.hasNext()) {
+			Object expectedElement = expectedIterator.next();
+			Object actualElement = actualIterator.next();
+
+			if (expectedElement == actualElement) {
+				processed++;
+				continue;
+			}
+
+			indexes.addLast(processed++);
+			assertIterableElementsEqual(expectedElement, actualElement, indexes, messageSupplier);
+			indexes.removeLast();
+		}
+
+		assertIteratorsAreEmpty(expectedIterator, actualIterator, processed, indexes, messageSupplier);
+	}
+
+	private static void assertIterableElementsEqual(Object expected, Object actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+		if (expected instanceof Iterable && actual instanceof Iterable) {
+			assertIterableEquals((Iterable<?>) expected, (Iterable<?>) actual, indexes, messageSupplier);
+		}
+		else if (!Objects.equals(expected, actual)) {
+			if (expected == null) {
```

Comment:
Now this really is a duplication of the method `assertIterablesNotNull`. Please replace with:

```
    assertIterablesNotNull(expected, actual, indexes, messageSupplier);
    failIterablesNotEqual(expected, actual, indexes, messageSupplier);
```

This will make this block very expressive. Thanks.